# It is a Regression Problem

## We are doing machine on House Price Dataset.

<table align="center" width=100%>
   <tr>
       </td>
          </td>
      </td>
      <td>
          <div align="center">
              <font color="#21618C" size=8px>
                  <b> Regression modeling on House Price Data
                  </b>
               </font>
            </div>
         </td>
      </tr>
 </table>   

# Problem Statement

This dataset is created for the prediction of House Prices from an Indian perspective Towns. The dataset contains various features that are important during the application for the selling the Houses. The predicted output obtained from the classification algorithm gives a fair idea about the chances of a Houses Sold.

In [1]:
#import 'Pandas'
#Pandas is utilize to format the data stucture.It means it can convert any data structure into pandas dataframe.
import pandas as pd

#import 'Numpy'
#We use Numpy to find statistical summary of Dataset
import numpy as np

#import subpackage of Matplotlib
import matplotlib.pyplot as plt

#import 'seaborn'
import seaborn as sns

# to suppress warnings
from warnings import filterwarnings
filterwarnings('ignore')


### Load the Dataset

In [2]:
# Load the data and name it 'df_house_price'
df_houseprice = pd.read_csv("House_Price1.csv")

# display first five observations using head()
df_houseprice.head()


price  resid_area  air_qual  room_num   age  distance  teachers  poor_prop  \
0   24.0       32.31     0.538     6.575  65.2      4.35      24.7       4.98   
1   21.6       37.07     0.469     6.421  78.9      4.99      22.2       9.14   
2   34.7       37.07     0.469     7.185  61.1      5.03      22.2       4.03   
3   33.4       32.18     0.458     6.998  45.8      6.21      21.3       2.94   
4   36.2       32.18     0.458     7.147  54.2      6.16      21.3       5.33   

  airport  n_hos_beds  n_hot_rooms waterbody  rainfall bus_ter     parks  Sold  
0     YES       5.480      11.1920     River        23     YES  0.049347     0  
1      NO       7.332      12.1728      Lake        42     YES  0.046146     1  
2      NO       7.394     101.1200       NaN        38     YES  0.045764     0  
3     YES       9.268      11.2672      Lake        45     YES  0.047151     0  
4      NO       8.824      11.2896      Lake        55     YES  0.039474     0

In [3]:
# display last five observations using head()
df_houseprice.tail()


price  resid_area  air_qual  room_num   age  distance  teachers  \
501   22.4       41.93     0.573     6.593  69.1      2.64      19.0   
502   20.6       41.93     0.573     6.120  76.7      2.44      19.0   
503   23.9       41.93     0.573     6.976  91.0      2.34      19.0   
504   22.0       41.93     0.573     6.794  89.3      2.54      19.0   
505   19.0       41.93     0.573     6.030  80.8      2.72      19.0   

     poor_prop airport  n_hos_beds  n_hot_rooms       waterbody  rainfall  \
501       9.67      NO       9.348      12.1792  Lake and River        27   
502       9.08     YES       6.612      13.1648  Lake and River        20   
503       5.64      NO       5.478      12.1912             NaN        31   
504       6.48     YES       7.940      15.1760             NaN        47   
505       7.88     YES      10.280      10.1520             NaN        45   

    bus_ter     parks  Sold  
501     YES  0.056006     1  
502     YES  0.059903     1  
503     YES  0.057572     1  
504     YES  0.060694     1  
505     YES  0.060336     1

## Data Overview

In [4]:
# Name the Columns
df_houseprice.columns

Index(['price', 'resid_area', 'air_qual', 'room_num', 'age', 'distance',
       'teachers', 'poor_prop', 'airport', 'n_hos_beds', 'n_hot_rooms',
       'waterbody', 'rainfall', 'bus_ter', 'parks', 'Sold'],
      dtype='object')

**Interpretation**

 - `Price`: Price of the House.
 - `resid_area` : Residential Area where my House is located.

In [5]:
# Basic Info of the Dataset
df_houseprice.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 16 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   price        506 non-null    float64
 1   resid_area   506 non-null    float64
 2   air_qual     506 non-null    float64
 3   room_num     506 non-null    float64
 4   age          506 non-null    float64
 5   distance     506 non-null    float64
 6   teachers     506 non-null    float64
 7   poor_prop    506 non-null    float64
 8   airport      506 non-null    object 
 9   n_hos_beds   498 non-null    float64
 10  n_hot_rooms  506 non-null    float64
 11  waterbody    351 non-null    object 
 12  rainfall     506 non-null    int64  
 13  bus_ter      506 non-null    object 
 14  parks        506 non-null    float64
 15  Sold         506 non-null    int64  
dtypes: float64(11), int64(2), object(3)
memory usage: 63.4+ KB


**Interpretation**
 - We have 13 Numerical colums and 3 Categorical columns.
 - We have null values in the data.
 - We have total 16 columns in the data

In [6]:
# Shape of the Data
df_houseprice.shape


(506, 16)

**Interpretation**
 - We have total 16 Columns.
 - We have 506 observations.
    

In [7]:
# Summarize the basic statistics of the data.
df_houseprice.describe().T

count       mean        std        min        25%        50%  \
price        506.0  22.528854   9.182176   5.000000  17.025000  21.200000   
resid_area   506.0  41.136779   6.860353  30.460000  35.190000  39.690000   
air_qual     506.0   0.554695   0.115878   0.385000   0.449000   0.538000   
room_num     506.0   6.284634   0.702617   3.561000   5.885500   6.208500   
age          506.0  68.574901  28.148861   2.900000  45.025000  77.500000   
distance     506.0   3.971996   2.108532   1.130000   2.270000   3.385000   
teachers     506.0  21.544466   2.164946  18.000000  19.800000  20.950000   
poor_prop    506.0  12.653063   7.141062   1.730000   6.950000  11.360000   
n_hos_beds   498.0   7.899767   1.476683   5.268000   6.634500   7.999000   
n_hot_rooms  506.0  13.041605   5.238957  10.057600  11.189800  12.720000   
rainfall     506.0  39.181818  12.513697   3.000000  28.000000  39.000000   
parks        506.0   0.054454   0.010632   0.033292   0.046464   0.053507   
Sold         506.0   0.454545   0.498422   0.000000   0.000000   0.000000   

                   75%         max  
price        25.000000   50.000000  
resid_area   48.100000   57.740000  
air_qual      0.624000    0.871000  
room_num      6.623500    8.780000  
age          94.075000  100.000000  
distance      5.367500   12.320000  
teachers     22.600000   27.400000  
poor_prop    16.955000   37.970000  
n_hos_beds    9.088000   10.876000  
n_hot_rooms  14.170800  101.120000  
rainfall     50.000000   60.000000  
parks         0.061397    0.086711  
Sold          1.000000    1.000000

**Intrepretation**
 - We have Outliers in the data, most of the outliers are in AGE column.

## Data Preprocessing 

### a. Data Cleaning

In [8]:
#show the 1st five observations
df_houseprice.head()

price  resid_area  air_qual  room_num   age  distance  teachers  poor_prop  \
0   24.0       32.31     0.538     6.575  65.2      4.35      24.7       4.98   
1   21.6       37.07     0.469     6.421  78.9      4.99      22.2       9.14   
2   34.7       37.07     0.469     7.185  61.1      5.03      22.2       4.03   
3   33.4       32.18     0.458     6.998  45.8      6.21      21.3       2.94   
4   36.2       32.18     0.458     7.147  54.2      6.16      21.3       5.33   

  airport  n_hos_beds  n_hot_rooms waterbody  rainfall bus_ter     parks  Sold  
0     YES       5.480      11.1920     River        23     YES  0.049347     0  
1      NO       7.332      12.1728      Lake        42     YES  0.046146     1  
2      NO       7.394     101.1200       NaN        38     YES  0.045764     0  
3     YES       9.268      11.2672      Lake        45     YES  0.047151     0  
4      NO       8.824      11.2896      Lake        55     YES  0.039474     0

In [9]:
#Rename the columns 
#Step1: Show the column names
df_houseprice.columns

Index(['price', 'resid_area', 'air_qual', 'room_num', 'age', 'distance',
       'teachers', 'poor_prop', 'airport', 'n_hos_beds', 'n_hot_rooms',
       'waterbody', 'rainfall', 'bus_ter', 'parks', 'Sold'],
      dtype='object')

In [10]:
#Step2: Allocate the new column names to old column name
#Make the dictionary {'old column name'; 'new column name'}
d1={'price':'Price',
   'resid_area':'Residential_Area',
   'air_qual':'Air_quality',
   'room_num':'Room_number',
   'age': 'Age',
   'distance': 'Distance',
   'teachers': 'Teachers',
   'poor_prop':'Poor_proportion',
   'airport':'Airport',
   'n_hos_beds':'n_Hospital_Beds',
   'n_hot_rooms': 'n_Hotel_rooms',
   'waterbody':'Waterbody',
   'rainfall':'Rainfall',
   'bus_ter': 'Bus_terminal',
   'parks':'Parks',
   'Sold':'Sold',
   }


In [11]:
df_houseprice = df_houseprice.rename(columns=d1)
df_houseprice

Price  Residential_Area  Air_quality  Room_number   Age  Distance  \
0     24.0             32.31        0.538        6.575  65.2      4.35   
1     21.6             37.07        0.469        6.421  78.9      4.99   
2     34.7             37.07        0.469        7.185  61.1      5.03   
3     33.4             32.18        0.458        6.998  45.8      6.21   
4     36.2             32.18        0.458        7.147  54.2      6.16   
..     ...               ...          ...          ...   ...       ...   
501   22.4             41.93        0.573        6.593  69.1      2.64   
502   20.6             41.93        0.573        6.120  76.7      2.44   
503   23.9             41.93        0.573        6.976  91.0      2.34   
504   22.0             41.93        0.573        6.794  89.3      2.54   
505   19.0             41.93        0.573        6.030  80.8      2.72   

     Teachers  Poor_proportion Airport  n_Hospital_Beds  n_Hotel_rooms  \
0        24.7             4.98     YES            5.480        11.1920   
1        22.2             9.14      NO            7.332        12.1728   
2        22.2             4.03      NO            7.394       101.1200   
3        21.3             2.94     YES            9.268        11.2672   
4        21.3             5.33      NO            8.824        11.2896   
..        ...              ...     ...              ...            ...   
501      19.0             9.67      NO            9.348        12.1792   
502      19.0             9.08     YES            6.612        13.1648   
503      19.0             5.64      NO            5.478        12.1912   
504      19.0             6.48     YES            7.940        15.1760   
505      19.0             7.88     YES           10.280        10.1520   

          Waterbody  Rainfall Bus_terminal     Parks  Sold  
0             River        23          YES  0.049347     0  
1              Lake        42          YES  0.046146     1  
2               NaN        38          YES  0.045764     0  
3              Lake        45          YES  0.047151     0  
4              Lake        55          YES  0.039474     0  
..              ...       ...          ...       ...   ...  
501  Lake and River        27          YES  0.056006     1  
502  Lake and River        20          YES  0.059903     1  
503             NaN        31          YES  0.057572     1  
504             NaN        47          YES  0.060694     1  
505             NaN        45          YES  0.060336     1  

[506 rows x 16 columns]

In [12]:
# Authenticate the Data
df_houseprice.columns

Index(['Price', 'Residential_Area', 'Air_quality', 'Room_number', 'Age',
       'Distance', 'Teachers', 'Poor_proportion', 'Airport', 'n_Hospital_Beds',
       'n_Hotel_rooms', 'Waterbody', 'Rainfall', 'Bus_terminal', 'Parks',
       'Sold'],
      dtype='object')

### b. Null Value handling

In [13]:
# Find the valid null values 
df_houseprice.isnull().sum()

Price                 0
Residential_Area      0
Air_quality           0
Room_number           0
Age                   0
Distance              0
Teachers              0
Poor_proportion       0
Airport               0
n_Hospital_Beds       8
n_Hotel_rooms         0
Waterbody           155
Rainfall              0
Bus_terminal          0
Parks                 0
Sold                  0
dtype: int64

In [14]:
# Percentage missing values
round((df_houseprice.isnull().sum())/len(df_houseprice)*100,3)

Price                0.000
Residential_Area     0.000
Air_quality          0.000
Room_number          0.000
Age                  0.000
Distance             0.000
Teachers             0.000
Poor_proportion      0.000
Airport              0.000
n_Hospital_Beds      1.581
n_Hotel_rooms        0.000
Waterbody           30.632
Rainfall             0.000
Bus_terminal         0.000
Parks                0.000
Sold                 0.000
dtype: float64

In [15]:
# Find the nature of the column
df_houseprice.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 16 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Price             506 non-null    float64
 1   Residential_Area  506 non-null    float64
 2   Air_quality       506 non-null    float64
 3   Room_number       506 non-null    float64
 4   Age               506 non-null    float64
 5   Distance          506 non-null    float64
 6   Teachers          506 non-null    float64
 7   Poor_proportion   506 non-null    float64
 8   Airport           506 non-null    object 
 9   n_Hospital_Beds   498 non-null    float64
 10  n_Hotel_rooms     506 non-null    float64
 11  Waterbody         351 non-null    object 
 12  Rainfall          506 non-null    int64  
 13  Bus_terminal      506 non-null    object 
 14  Parks             506 non-null    float64
 15  Sold              506 non-null    int64  
dtypes: float64(11), int64(2), object(3)
memory u

In [16]:
# This is a Categorical column so we can replace the null with Mode.
df_houseprice.Waterbody.mode()

0    River
Name: Waterbody, dtype: object

In [17]:
# Replace the null in that column
df_houseprice.wWterbody=df_houseprice.Waterbody.fillna('river')

In [18]:
# Percentage missing values
round((df_houseprice.isnull().sum())/len(df_houseprice)*100,3)

Price                0.000
Residential_Area     0.000
Air_quality          0.000
Room_number          0.000
Age                  0.000
Distance             0.000
Teachers             0.000
Poor_proportion      0.000
Airport              0.000
n_Hospital_Beds      1.581
n_Hotel_rooms        0.000
Waterbody           30.632
Rainfall             0.000
Bus_terminal         0.000
Parks                0.000
Sold                 0.000
dtype: float64

In [19]:
# In n_Hospital_Beds contain less than 10% values so we can drop the null values.
df_houseprice.dropna(inplace=True)

In [20]:
# Percentage missing values
round((df_houseprice.isnull().sum())/len(df_houseprice)*100,3)

Price               0.0
Residential_Area    0.0
Air_quality         0.0
Room_number         0.0
Age                 0.0
Distance            0.0
Teachers            0.0
Poor_proportion     0.0
Airport             0.0
n_Hospital_Beds     0.0
n_Hotel_rooms       0.0
Waterbody           0.0
Rainfall            0.0
Bus_terminal        0.0
Parks               0.0
Sold                0.0
dtype: float64

In [21]:
## Prepare data for Machine Learning
# We know that the machine learning algorithms are mathematical expressions, so we need to provide the data to the algorithm in specific format
# Encoding
# Step-1 : Divide the data based on their datatypes
df_cat = df_houseprice.select_dtypes(include=['object'])
df_cat.columns


Index(['Airport', 'Waterbody', 'Bus_terminal'], dtype='object')

In [22]:
# Use Dummy Encoding
dummy_var =pd.get_dummies(data = df_cat, drop_first = True)
dummy_var = dummy_var.reset_index()

In [23]:
# Use Dummy Encoding
dummy_var2 =pd.get_dummies(data = df_cat)
dummy_var2

Airport_NO  Airport_YES  Waterbody_Lake  Waterbody_Lake and River  \
0         False         True           False                     False   
1          True        False            True                     False   
3         False         True            True                     False   
4          True        False            True                     False   
6         False         True           False                     False   
..          ...          ...             ...                       ...   
498       False         True           False                     False   
499        True        False            True                     False   
500        True        False            True                     False   
501        True        False           False                      True   
502       False         True           False                      True   

     Waterbody_River  Bus_terminal_YES  
0               True              True  
1              False              True  
3              False              True  
4              False              True  
6               True              True  
..               ...               ...  
498             True              True  
499            False              True  
500            False              True  
501            False              True  
502            False              True  

[343 rows x 6 columns]

In [24]:
df_num = df_houseprice.select_dtypes(include = [np.number])
df_num.head()

Price  Residential_Area  Air_quality  Room_number   Age  Distance  \
0   24.0             32.31        0.538        6.575  65.2      4.35   
1   21.6             37.07        0.469        6.421  78.9      4.99   
3   33.4             32.18        0.458        6.998  45.8      6.21   
4   36.2             32.18        0.458        7.147  54.2      6.16   
6   22.9             37.87        0.524        6.012  66.6      5.87   

   Teachers  Poor_proportion  n_Hospital_Beds  n_Hotel_rooms  Rainfall  \
0      24.7             4.98            5.480        11.1920        23   
1      22.2             9.14            7.332        12.1728        42   
3      21.3             2.94            9.268        11.2672        45   
4      21.3             5.33            8.824        11.2896        55   
6      24.8            12.43            6.958        12.1832        41   

      Parks  Sold  
0  0.049347     0  
1  0.046146     1  
3  0.047151     0  
4  0.039474     0  
6  0.052170     1

In [25]:
from sklearn.preprocessing import StandardScaler

In [26]:
df_target = df_num['Price']
df_num = df_num.drop('Price', axis = 1)

In [27]:
# Initiate the StandardScaler
ss = StandardScaler()

In [28]:
# Transform the Data
Scale = ss.fit_transform(df_num)

In [29]:
Scale

array([[-1.22113906, -0.08708422,  0.35127837, ..., -1.33138234,
        -0.42318421, -0.92415045],
       [-0.52970502, -0.68208471,  0.14419804, ...,  0.19056446,
        -0.72689187,  1.08207489],
       [-1.24002276, -0.77693986,  0.92007695, ...,  0.43087185,
        -0.63156184, -0.92415045],
       ...,
       [-0.14912577,  0.31820597, -0.38560489, ...,  1.15179402,
        -0.12437127, -0.92415045],
       [ 0.17625495,  0.21472763,  0.37548257, ..., -1.01097249,
         0.20842292,  1.08207489],
       [ 0.17625495,  0.21472763, -0.26054989, ..., -1.57168973,
         0.57815884,  1.08207489]])

In [30]:
df_Scale = pd.DataFrame(Scale, columns = df_num.columns)
df_Scale = df_Scale.reset_index()

In [31]:
df_Scale.head()

index  Residential_Area  Air_quality  Room_number       Age  Distance  \
0      0         -1.221139    -0.087084     0.351278 -0.049985  0.139406   
1      1         -0.529705    -0.682085     0.144198  0.425106  0.441920   
2      2         -1.240023    -0.776940     0.920077 -0.722742  1.018587   
3      3         -1.240023    -0.776940     1.120434 -0.431445  0.994953   
4      4         -0.413498    -0.207809    -0.405775 -0.001436  0.857877   

   Teachers  Poor_proportion  n_Hospital_Beds  n_Hotel_rooms  Rainfall  \
0  1.384755        -1.015715        -1.606757      -0.429601 -1.331382   
1  0.233366        -0.423275        -0.378839      -0.186908  0.190564   
2 -0.181134        -1.306238         0.904773      -0.410994  0.430872   
3 -0.181134        -0.965870         0.610391      -0.405451  1.231896   
4  1.430811         0.045265        -0.626810      -0.184335  0.110462   

      Parks      Sold  
0 -0.423184 -0.924150  
1 -0.726892  1.082075  
2 -0.631562 -0.924150  
3 -1.359756 -0.924150  
4 -0.155435  1.082075

In [32]:
# Make a DataFrame for independent variables
df_independent = pd.concat([df_Scale, dummy_var], axis = 1)
df_independent.head()

index  Residential_Area  Air_quality  Room_number       Age  Distance  \
0      0         -1.221139    -0.087084     0.351278 -0.049985  0.139406   
1      1         -0.529705    -0.682085     0.144198  0.425106  0.441920   
2      2         -1.240023    -0.776940     0.920077 -0.722742  1.018587   
3      3         -1.240023    -0.776940     1.120434 -0.431445  0.994953   
4      4         -0.413498    -0.207809    -0.405775 -0.001436  0.857877   

   Teachers  Poor_proportion  n_Hospital_Beds  n_Hotel_rooms  Rainfall  \
0  1.384755        -1.015715        -1.606757      -0.429601 -1.331382   
1  0.233366        -0.423275        -0.378839      -0.186908  0.190564   
2 -0.181134        -1.306238         0.904773      -0.410994  0.430872   
3 -0.181134        -0.965870         0.610391      -0.405451  1.231896   
4  1.430811         0.045265        -0.626810      -0.184335  0.110462   

      Parks      Sold  index  Airport_YES  Waterbody_Lake and River  \
0 -0.423184 -0.924150      0         True                     False   
1 -0.726892  1.082075      1        False                     False   
2 -0.631562 -0.924150      3         True                     False   
3 -1.359756 -0.924150      4        False                     False   
4 -0.155435  1.082075      6         True                     False   

   Waterbody_River  
0             True  
1            False  
2            False  
3            False  
4             True

In [33]:
import statsmodels
import statsmodels.api as sm
df_independent = sm.add_constant(df_independent)

In [34]:
from sklearn.model_selection import train_test_split

In [36]:
X_train, X_test, Y_train, Y_test = train_test_split(df_independent,
                                                   df_target,
                                                  random_state = 10,
                                                  test_size = 0.2)

In [38]:
print('X_train', X_train.shape)
print('Y_train', Y_train.shape)

print('X_test', X_test.shape)
print('Y_test', Y_test.shape)


X_train (274, 18)
Y_train (274,)
X_test (69, 18)
Y_test (69,)


# Linear Regression

In [39]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

In [43]:
model = LinearRegression()
model.fit(X_train, Y_train)
predictions = model.predict(X_test)
mse = mean_squared_error(Y_test, predictions)
print("Mean Squared Error:",mse)

Mean Squared Error: 25.98165559218157


In [44]:
predictions


array([27.65597303, 25.21337915, 18.17381059, 20.63248951, 17.62592603,
       19.84613733, 17.96959662, 32.06328268, 22.93669867, 36.23354952,
       15.6922288 , 42.78521807, 38.38781798, 25.42702213, 31.67514018,
       37.5410411 , 31.31732029, 25.40855914, 20.40083202, 22.51213133,
       19.50363371, 25.6590486 , 18.19221133, 23.64864189, 24.26816071,
       21.29168338, 17.87888963, 29.66591594, 16.72529114, 25.8644227 ,
       27.53682489, 28.79525716, 24.96776441, 28.30942553, 32.08785785,
       34.11524281, 10.72732232, 30.43419679, 19.81823965, 19.34613349,
       21.15725358, 23.06813308, 24.80030649, 20.43229413, 12.54220409,
       31.31938977, 19.67062666, 22.11195516, 17.5352817 , 22.69917288,
       24.41539796, 14.48785157, 19.7504597 , 15.44063443, 19.45020651,
       20.68767696, 25.65547455, 28.03286462, 18.92126559, 21.6759836 ,
       41.78746086, 38.37530777, 26.55274816, 15.00184886, 42.78107899,
       20.18871728, 15.31571316, 12.64904725, 21.38804275])

In [50]:
from sklearn.metrics import mean_absolute_error
mae = mean_absolute_error(Y_test, predictions)
print("Mean Absolute Error:",mae)

Mean Absolute Error: 3.6593447642702808


In [52]:
rms = mse**0.5
print("root_mean_square",rms)

root_mean_square 5.097220379008697
